ws/client.go

```diff
@@ -0,0 +1,175 @@
+package ws
+
+import (
+	"bytes"
+	"encoding/json"
+	"net/http"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/textileio/go-textile-core/thread"
+)
+
+const (
+	// Time allowed to write a message to the peer.
+	writeWait = 10 * time.Second
+
+	// Time allowed to read the next pong message from the peer.
+	pongWait = 60 * time.Second
+
+	// Send pings to peer with this period. Must be less than pongWait.
+	pingPeriod = (pongWait * 9) / 10
+
+	// Maximum message size allowed from peer.
+	maxMessageSize = 512
+)
+
+var (
+	upgrader = websocket.Upgrader{
+		ReadBufferSize:  1024,
+		WriteBufferSize: 1024,
+		CheckOrigin: func(r *http.Request) bool {
+			// @todo: auth with follow and/or read key
+			return true
+		},
+	}
+
+	newline = []byte{'\n'}
+	space   = []byte{' '}
+)
+
+type msg struct {
+	Type    string   `json:"type"`
+	Threads []string `json:"threads"`
+}
+
+// Client is a middleman between the websocket connection and the hub.
+type Client struct {
+	hub *Hub
+
+	// The websocket connection.
+	conn *websocket.Conn
+
+	// Buffered channel of outbound messages.
+	send chan []byte
+
+	// Active threads.
+	threads map[thread.ID]struct{}
+}
+
+// readPump pumps messages from the websocket connection to the hub.
+//
+// The application runs readPump in a per-connection goroutine. The application
+// ensures that there is at most one reader on a connection by executing all
+// reads from this goroutine.
+func (c *Client) readPump() {
+	defer func() {
+		c.hub.unregister <- c
+		c.conn.Close()
+	}()
+	c.conn.SetReadLimit(maxMessageSize)
+	_ = c.conn.SetReadDeadline(time.Now().Add(pongWait))
+	c.conn.SetPongHandler(func(string) error {
+		_ = c.conn.SetReadDeadline(time.Now().Add(pongWait))
+		return nil
+	})
+	for {
+		_, message, err := c.conn.ReadMessage()
+		if err != nil {
+			if websocket.IsUnexpectedCloseError(
+				err,
+				websocket.CloseGoingAway,
+				websocket.CloseAbnormalClosure) {
+				log.Errorf("error reading message: %s", err)
+			}
+			break
+		}
+		message = bytes.TrimSpace(bytes.Replace(message, newline, space, -1))
+
+		var m msg
+		err = json.Unmarshal(message, &m)
+		if err != nil {
+			log.Errorf("error unmarshaling message: %s", err)
+			break
+		}
+
+		switch m.Type {
+		case "subscribe":
+			for _, t := range m.Threads {
+				id, err := thread.Decode(t)
+				if err == nil {
+					log.Debugf("client requested thread %s", id.String())
+
+					c.threads[id] = struct{}{}
+				}
+			}
+		}
+	}
+}
+
+// writePump pumps messages from the hub to the websocket connection.
+//
+// A goroutine running writePump is started for each connection. The
+// application ensures that there is at most one writer to a connection by
+// executing all writes from this goroutine.
+func (c *Client) writePump() {
```

Comment:
Most `.Write` errors ignored since is a PoC? Maybe an extra `// ToDo` to keep track?

Reply:
Added!